server/middleware/featureGate.test.ts

```diff
@@ -1,43 +1,95 @@
+  } as unknown as Response
+}
+
+/**
+ * Asserts that the handler successfully responds
+ */
+function expectRequestHandlerToBeCalled(handler: RequestHandler, req: Request, res: Response): void {
+  const next: jest.Mock<NextFunction> = jest.fn()
+  handler(req, res, next)
+  expect(res.send).toBeCalledWith('OK')
+  expect(next).not.toBeCalled()
+}
+
+/**
+ * Asserts that the handler is not called and a 404 is passed to the next handler
+ */
+function expectRequestHandlerTo404(handler: RequestHandler, req: Request, res: Response): void {
+  const next: jest.Mock<NextFunction> = jest.fn()
+  handler(req, res, next)
+  expect(res.send).not.toBeCalled()
+  expect(next).toBeCalled()
+  expect(next.mock.calls[0][0].status).toEqual(404)
+}
 
 describe('featureGate', () => {
-  const gatedHandler: RequestHandler = featureGate('testFlag', (req, res) => {
-    res.send('OK')
-  })
+  /** Gated request handler that requires flag "testFlag" */
+  const gatedHandler = featureGate('testFlag', simpleRequestHandler)
 
   it('calls handler when feature is turned on', () => {
-    const req = {
-      app: {
-        locals: {
-          featureFlags: {
-            testFlag: true,
-          },
-        },
-      },
-    } as unknown as Request
-    const res = { send: jest.fn() } as unknown as Response
-    const next: jest.Mock<NextFunction> = jest.fn()
-    gatedHandler(req, res, next)
-    expect(res.send).toBeCalledWith('OK')
-    expect(next).not.toBeCalled()
+    const req = mockRequest({ featureFlags: { testFlag: true } })
+    const res = mockResponse()
+    expectRequestHandlerToBeCalled(gatedHandler, req, res)
   })
 
   it('returns 404 when feature is turned off', () => {
-    const req = {
-      app: {
-        locals: {
-          featureFlags: {
-            testFlag: false,
-          },
-        },
-      },
-    } as unknown as Request
-    const res = { send: jest.fn() } as unknown as Response
-    const next: jest.Mock<NextFunction> = jest.fn()
-    gatedHandler(req, res, next)
-    expect(res.send).not.toBeCalled()
-    expect(next).toBeCalled()
-    expect(next.mock.calls[0][0].status).toEqual(404)
+    const req = mockRequest({ featureFlags: { testFlag: false } })
+    const res = mockResponse()
+    expectRequestHandlerTo404(gatedHandler, req, res)
+  })
+})
+
+describe('caseloadGate', () => {
+  /** Gated request handler that requires case load to include BWI or MDI */
+  const gatedHandler = caseloadGate(['BWI', 'MDI'], simpleRequestHandler)
+
+  it('calls handler when user’s active case load is included in specified prisons', () => {
+    const req = mockRequest()
+    const res = mockResponse({ user: { activeCaseload: { id: 'MDI' } } })
+    expectRequestHandlerToBeCalled(gatedHandler, req, res)
+  })
+
+  it('calls handler when user’s case loads overlap with specified prisons', () => {
+    const req = mockRequest()
+    const res = mockResponse({ user: { activeCaseload: { id: 'LEI' }, caseloads: [{ id: 'LEI' }, { id: 'BWI' }] } })
```

Comment:
Maybe it's explained later in the diff but I'm confused by the fact this wouldn't receive a `404` response given the graph would still use data from `LEI` but that prison doesn't have the flag enabled (in this case it's enabled only for BWI/MDI).

Reply:
I know see there are two separate middleware with different behaviour 👌